### DISCUSSION TITLE
Social plugin: runtime error with virtual files

### DISCUSSION THREAD

**Discussion by thorstenpeter-sophos** [squidfunk/mkdocs-material#7517] [General]
This is actually a bug report. I didn't open an issue because I can't create a minimal reproduction. The bug only manifests in combination with other plugins that add virtual files to the build. I hope you still consider this report as valid.

The Insiders version of the material/social plugin has the following code in its on_files event handler:

```
for file in files:
    if file.abs_src_path.startswith(_templates_dirpath()):
        file.inclusion = InclusionLevel.EXCLUDED
```

This code raises an AttributeError exception if abs_src_path is None. This happens for virtual files, which MkDocs supports since version 1.6.

See for example here:
https://www.mkdocs.org/about/release-notes/#enabling-true-generated-files-and-expanding-the-file-api

    Further, abs_src_path is no longer guaranteed to be present and can be None instead. MkDocs itself still uses physical files in all cases, but eventually plugins will appear that don't populate this attribute.

Our internal MkDocs plugin creates virtual files in its on_files event handler. I was able to work around the issue in the Social plugin by adjusting the event handler priorities to ensure that we create these files after the Social on_files event has run. However, not all third-party plugins might be able to make similar adjustments.

**Comment by squidfunk** [comment]
Thanks for noting. Unfortunately we need a minimal reproduction to fix it, which we'd be more than happy to. Thus, if you can provide one, it's likely an easy fix. You might just put the virtual files logic in a hook to reproduce it, but I'm not sure since I don't know the code. Without a reproduction, we can only guess.

**Comment by thorstenpeter-sophos** [comment]
I managed to prepare a reproduction by creating a minimal example plugin that adds a generated file to the build.
Here's the issue: https://github.com/squidfunk/mkdocs-material/issues/7520

**Comment by thorstenpeter-sophos** [comment]
Thanks a lot for fixing this so quickly in [ee8d243f9](https://github.com/squidfunk/mkdocs-material-insiders/commit/ee8d243f989a46ff8fe391c5f8245650dede8839).

And thank you for reminding me of the hooks feature, which I wasn't aware of. This might be helpful for future bug reports.
